docs/docs/dev_docs/contracts/state_variables.md

```diff
@@ -1,35 +1,220 @@
+
+- [Singleton<NoteType\>](#singletonnotetype)
+- [ImmutableSingleton<NoteType\>](#immutablesingletonnotetype)
+- [Set<NoteType\>](#setnotetype)
+
+In contrast to public state variables, private state variables are only visible to specific relevant parties. The value of a private state variable can either be shared via (log)[INSERT_LINK_HERE] or completely offline.
+
+Note that an app can choose to emit the data via unencrypted log, or define a note whose data is easy to figure out, then the information is technically not private and could be visible to anyone.
 
 ### Notes
 
-### Custom Notes
+Unlike public state variables, which can be arbitrary types. Private state variables operate on `NoteType`.
+
+Notes are the fundamental elements in the private world.
+
+A note should confine to the following interface:
+
+#include_code NoteInterface /yarn-project/noir-libs/noir-aztec/src/note/note_interface.nr rust
+
+## `Singleton<NoteType>`
+
+Singleton is a private state variable that is unique in a way. When a singleton is initialised, a note is created to represent its value. And the way to update the value is to destroy the current note, and create a new one with the updated value.
+
+### `::new`
+
+Here we define a singleton for storing a Card note:
+
+#include_code state_vars-Singleton /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/storage/legendary_card.nr rust
+
+### `.initialise`
+
+The initial value of a singleton is set via calling `initialise`:
+
+#include_code state_vars-SingletonInit /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/actions.nr rust
+
+When this function is called, a nullifier of the storage slot is created, preventing this singleton from being initialised again.
+
+### `.replace`
+
+To modify the value of a singleton, we will create a note (a Card in the following example) that has the new data, and replace the current note with it:
+
+#include_code state_vars-SingletonReplace /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/actions.nr rust
+
+This function will destroy the old note under the hood. If two people are trying to modify the singleton at the same time, only one will succeed.
+
+### `.get_note`
+
+This function allows us to get the note of a singleton:
+
+#include_code state_vars-SingletonGet /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/actions.nr rust
+
+However, it's possible that at the time this function is called, the system hasn't synched to the block where the latest note was created. Or a malicious user might feed an old state to this function, tricking the proving system into thinking that the value hasn't changed. To avoid an attack around it, this function will destroy the current note, and replace it with a duplicated note that has the same fields. Because the nullifier of the latest note will be emitted, if two people are trying to use this function against the same note, only one will succeed.
```

Comment:
`get_note` will destroy the current note and create a new note, and `replace` will destroy the new note and create the replacement note. So it should work just fine. Although the extra read request and nullifier are really annoying.
But I don't know if it's worth teaching people the differences of the apis to save 1 extra slot in MAX___PER_CALL. 

What if we need to call `replace` twice in a call:
```rust
let mut note = singleton.get_note_as_mutable();
for i in 0..4 {
  let score = contractB.play_round()
  if score > some_value {
    note.count += 1;
    var.replace(note);
  }
}
```
> I guess we can reset `_already_got` (maybe rename it to `_nullified`?) after the replacement note is created!

Or, what if a nested function also needs to access the singleton:
```rust
fn A() {
  let mut note = singleton.get_note_as_mutable();
  if note.owner == some_value {
    this.B()
  }
  note.count += 1;
  singleton.replace(note);
}

fn B() {
  let note = singleton.get_note_as_read_only();
  if note.owner == some_value {
    do_something();
  }
}
```
Function A will nullify the current note. And function B will also trying to nullify the current note (same nullifier!), and create a duplicated note. Then function A will create the replacement note.

Reply:
Sorry, I'm a bit delayed catching up.

> What if we need to call replace twice in a call:

That example is probably just something we should teach as a bad pattern. The same can be achieved more efficiently if `.replace` is called only at the end, instead of every iteration.

> Or, what if a nested function also needs to access the singleton:

In that example, the function `B` could be revised to take the note as an argument.

HERE'S A THOUGHT...

To help users, the `context` could track information about each read, nullification, and write, and only emit read_requests / nullifiers / new commitments at the time that `context.finalise()` is called (or at the time an external `call` is made). That would be cool.

Reply:
#1898